### ISSUE TITLE
Process packet error - closed

### ISSUE THREAD

**Issue by sshhawwnn** [emqx/emqx#2336]
#### Environment

- OS: CentOS 7.6
- Erlang/OTP: 
- EMQ: 3.0.1

#### Description

*A description of the issue*
Subscriber unexpectedly disconnect or got terminated and when checking the logs, I found these.
```

2019-03-19 15:36:37.640 [error] SUB001@172.22.0.7:64735 Process packet error - closed
2019-03-19 16:21:08.342 [error] SUB001@172.22.0.7:49864 Process packet error - closed
2019-03-19 17:55:38.650 [error] SUB001@172.22.0.7:50815 Process packet error - closed
2019-03-19 19:39:39.632 [error] SUB001@172.22.0.7:56092 Process packet error - closed
2019-03-19 21:04:40.589 [error] SUB001@172.22.0.7:58474 Process packet error - closed
2019-03-19 21:32:40.897 [error] SUB001@172.22.0.7:58807 Process packet error - closed
2019-03-19 21:57:41.143 [error] SUB001@172.22.0.7:59749 Process packet error - closed
2019-03-20 01:08:55.985 [error] SUB001 ** Generic server <0.13195.10> terminating
** Last message in was {dispatch,<<"x/y/z/a/b/c/SUB002">>,
                           {message,
                               <<0,5,132,124,68,11,183,165,165,154,66,0,21,244,
                                 0,5>>,
                               0,<<"SUB002">>,
                               #{dup => false,retain => false},
                               #{username => <<"SUB002">>},
                               <<"x/y/z/a/b/c/SUB002">>,
                               <<"{\"data_type\":0,\"timestamp\":\"2019-03-20T01:08:55+00:00\",\"data\":[{}]}">>,
                               {1553,44135,983013}}}
** When Server state == {state,15000,false,remote,<<"SUB001">>,<<"Testing">>,
                            <38588.4276.66>,undefined,1,0,
                            #{<<"x/y/z/a/b/c/f/data/G0">> =>
                                  #{nl => 0,qos => 2,rap => 0,rc => 0,rh => 0},
                              <<"x/y/z/a/b/c/e/data/G0">> =>
                                  #{nl => 0,qos => 2,rap => 0,rc => 0,rh => 0},
                              <<"x/y/z/a/b/c/SUB002">> =>
                                  #{nl => 0,qos => 2,rap => 0,rc => 0,
                                    rh => 0}},
                            false,
                            {emqx_inflight,0,{0,nil}},
                            20000,undefined,
                            {mqueue,true,0,0,0,none,infinity,{queue,[],[],0}},
                            #{},100,300000,
                            #Ref<0.2444105538.239075329.108518>,259200,
                            undefined,true,
                            #Ref<0.2444105538.239075329.108500>,
                            {emqx_gc,
                                #{cnt => {1000,1000},
                                  oct => {1048576,1048576}}},
                            {1551,889431,730907},
                            undefined,undefined}
** Reason for termination ==
** {{badmatch,true},
    [{emqx_session,dispatch,2,[{file,"src/emqx_session.erl"},{line,843}]},
     {emqx_session,handle_info,2,[{file,"src/emqx_session.erl"},{line,584}]},
     {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,637}]},
     {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,711}]},
     {proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]}
2019-03-20 01:08:56.000 [error] SUB001 crasher:
    initial call: emqx_session:init/1
    pid: <0.13195.10>
    registered_name: []
    exception error: no match of right hand side value true
      in function  emqx_session:dispatch/2 (src/emqx_session.erl, line 843)
      in call from emqx_session:handle_info/2 (src/emqx_session.erl, line 584)
      in call from gen_server:try_dispatch/4 (gen_server.erl, line 637)
      in call from gen_server:handle_msg/6 (gen_server.erl, line 711)
    ancestors: [emqx_session_sup,emqx_sm_sup,emqx_sup,<0.1902.0>]
    message_queue_len: 1
    messages: [{'EXIT',<38588.4276.66>,
                      {{badmatch,true},
                       [{emqx_session,dispatch,2,
                            [{file,"src/emqx_session.erl"},{line,843}]},
                        {emqx_session,handle_info,2,
                            [{file,"src/emqx_session.erl"},{line,584}]},
                        {gen_server,try_dispatch,4,
                            [{file,"gen_server.erl"},{line,637}]},
                        {gen_server,handle_msg,6,
                            [{file,"gen_server.erl"},{line,711}]},
                        {proc_lib,wake_up,3,
                            [{file,"proc_lib.erl"},{line,259}]}]}}]
    links: [<0.1951.0>]
    dictionary: [{'$logger_metadata$',#{client_id => <<"SUB001">>}},
                  {deliver_stats,5},
                  {force_shutdown_policy,
                      #{max_heap_size => 0,message_queue_len => 0}}]
    trap_exit: true
    status: running
    heap_size: 10958
    stack_size: 27
    reductions: 17093808
  neighbours:
```

**Comment by HJianBo** [comment]
Hi, @sshhawwnn  We have fixed this bug at the v3.1.beta.2 version. Thanks for your report!

**Comment by sshhawwnn** [comment]
Hi @HJianBo , the problem still persists in v3.1.beta.2.


```
2019-03-21 08:28:05.017 [error] IB001 ** Generic server <0.1871.0> terminating 
,
     {proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]}
,
                            false,
,
** Last message in was {dispatch,<<"sg/sj/24k/ifocus/data/DG001">>,
,
                           {message,
,
                               <<0,5,132,150,132,105,115,189,165,155,0,0,8,151,
,
                               #{dup => false,retain => false},
,
                                 0,2>>,
,
                               0,<<"DG001">>,
,
                               #{allow_publish => true,
,
                                 username => <<"DG001">>},
,
                               <<"sg/sj/24k/ifocus/data/DG001">>,
,
                               <<"{\"data_type\":0,\"timestamp\":\"2019-03-21T08:28:05+00:00\",\"data\":[{}]}">>,
,
                               {1553,156885,17534}}}
,
** When Server state == {state,15000,false,#Fun<emqx_session.2.44155652>,
,
                            <<"IB001">>,<<"Testing">>,<38799.2430.0>,
,
                            undefined,1,0,
,
                            #{<<"sg/sj/24k/ascent/data/G0">> =>
,
                                  #{nl => 0,qos => 2,rap => 0,rc => 2,rh => 0},
,
                              <<"sg/sj/24k/bellbox/data/G0">> =>
,
                              <<"sg/sj/24k/ifocus/data/DG001">> =>
,
                                  #{nl => 0,qos => 2,rap => 0,rc => 2,
,
                                    rh => 0}},
,
                            {emqx_inflight,0,{0,nil}},
,
                            {mqueue,true,0,0,0,none,infinity,{queue,[],[],0}},
,
                            20000,undefined,
,
                            #{},100,300000,
,
                            #Ref<0.2366959605.1109917697.38981>,259200,
,
                            undefined,true,
,
                            #Ref<0.2366959605.1109917697.38980>,
,
                            {emqx_gc,
,
                                #{cnt => {1000,1000},
,
                                  oct => {1048576,1048576}}},
,
                            {1553,156522,888997},
,
                            undefined,undefined}
,
** {{badmatch,true},
,
** Reason for termination == 
,
    [{emqx_session,batch_process,2,[{file,"src/emqx_session.erl"},{line,895}]},
,
     {emqx_session,handle_dispatch,2,
,
     {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,637}]},
,
                   [{file,"src/emqx_session.erl"},{line,810}]},
,
     {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,711}]},

```

**Comment by HJianBo** [comment]
Cool! @sshhawwnn Thanks a lot of reproducing for this. It is a BUG really. I have fixed this line and will be released at tomorrow!

**Comment by sshhawwnn** [comment]
@HJianBo ok, I will try again when the new version is released.

**Comment by sshhawwnn** [comment]
EMQ: v3.1-beta.3
```
2019-03-23 09:10:00.734 [error] supervisor: 'esockd_connection_sup - <0.2275.0>'
    errorContext: connection_crashed
    reason: {{badmatch,{error,closed}},
             [{emqx_connection,init,1,
                               [{file,"src/emqx_connection.erl"},{line,146}]},
              {proc_lib,init_p_do_apply,3,
                        [{file,"proc_lib.erl"},{line,249}]}]}
    offender: [{pid,<0.23717.0>},
               {name,connection},
               {mfargs,
                   {emqx_connection,start_link,
                       [[{max_conn_rate,500},
                         {active_n,100},
                         {zone,external}]]}}]
2019-03-23 09:10:06.734 [error] crasher:
    initial call: emqx_connection:init/1
    pid: <0.23724.0>
    registered_name: []
    exception error: no match of right hand side value {error,closed}
      in function  emqx_connection:init/1 (src/emqx_connection.erl, line 146)
    ancestors: [<0.2275.0>,<0.2274.0>,esockd_sup,<0.2038.0>]
    message_queue_len: 0
    messages: []
    links: [<0.2275.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 4185
    stack_size: 27
    reductions: 5669
  neighbours:
```

**Comment by Gilbert-Wong** [comment]
This bug has been fixed in 3.1 beta3 release, please check it.
